### ISSUE TITLE
Python 3 release - will be 0.7.0

### ISSUE THREAD

**Issue by mikepurvis** [catkin/catkin_tools#594]
Would be great to have a released version of catkin_tools which works well for Python 3. Myself and some colleagues have been using this from source in Py3 virtualenvs for a few weeks and not discovered anything horrific; are there any objections before we ask @wjwwood to pull the lever?

**Comment by timonegk** [comment]
I am also using a source build of catkin_tools with python3 without any problems and would appreciate a new release.

**Comment by sloretz** [comment]
A new release would be appreciated on Debian Buster and Ubuntu Focal. It looks like it depends on `python3-trollius`, but that package doesn't exist on those platforms. Looks like that has been fixed on master already in #574

```
Package: python3-catkin-tools
Priority: optional
Section: python
Installed-Size: 786
Maintainer: William Woodall <william@osrfoundation.org>
Architecture: all
Version: 0.4.5-1
Depends: python3-catkin-pkg (>= 0.2.9), python3-pkg-resources, python3-trollius, python3-yaml, python3:any (>= 3.2~), python3-setuptools
Conflicts: python-catkin-tools
Filename: pool/main/p/python3-catkin-tools/python3-catkin-tools_0.4.5-1_all.deb
SHA512: bb6d05238d8a18685c2370b70cbe0f95edf433044ae1f8c59c1fd96742e9557cc620acb1d8094c3ce43452576415075c86404866fecce59eac4d8e46bff2a26c
Size: 313732
SHA256: 6abf1d25c38cef549d3dc1c3abd3f927fd5d71633a904c70563ddc34a1931d6c
SHA1: 5d040daee76c9e90893c68641656651ecaa8772f
MD5sum: b36033eecbf8f1050bd6277ba0105c02
Description: Command line tools for working with catkin.
 Provides command line tools for working with catkin.
```

**Comment by timonegk** [comment]
I guess #597 should be closed first, I suggested a fix in #601.

**Comment by DLu** [comment]
:+1:

**Comment by tykurtz** [comment]
Cutting a new release would be greatly appreciated. Running `0.4.5` on python3  https://github.com/catkin/catkin_tools/issues/510 and https://github.com/catkin/catkin_tools/issues/558#issuecomment-537216485 are still open issues which were fixed by switching to `master`.

**Comment by nim65s** [comment]
PyPI release is also broken for Python 3.8, because trollius defines `async`, which is now a reserved keyword:
```
Traceback (most recent call last):
  File "/home/nim/.local/bin/catkin", line 5, in <module>
    from catkin_tools.commands.catkin import main
  File "/home/nim/.local/lib/python3.8/site-packages/catkin_tools/commands/catkin.py", line 28, in <module>
    from catkin_tools.common import is_tty
  File "/home/nim/.local/lib/python3.8/site-packages/catkin_tools/common.py", line 23, in <module>
    import trollius as asyncio
  File "/home/nim/.local/lib/python3.8/site-packages/trollius/__init__.py", line 21, in <module>
    from .base_events import *
  File "/home/nim/.local/lib/python3.8/site-packages/trollius/base_events.py", line 42, in <module>
    from . import tasks
  File "/home/nim/.local/lib/python3.8/site-packages/trollius/tasks.py", line 565
    def async(coro_or_future, loop=None):
        ^
SyntaxError: invalid syntax
```

**Comment by facontidavide** [comment]
It would be great to get the new release on all the LTS version of ROS ;)
There are some recent changes that affect me, soooo :wink:

**Comment by davetcoleman** [comment]
This is blocking MoveIt releasing into Noetic :-/

**Comment by gavanderhoorn** [comment]
While a new release would be nice, you should be able to install `catkin_tools` using `pip3`. See my comment here: https://github.com/catkin/catkin_tools/issues/603#issuecomment-634617492 and the linked ROS Answers question.

**Comment by dirk-thomas** [comment]
> This is blocking MoveIt releasing into Noetic :-/

Out of curiosity: what feature of `catkin_tools` does MoveIt (I assume the CI?) need which can't be satisfied by `catkin_make_isolated` or perhaps `colcon`?

**Comment by dirk-thomas** [comment]
I was just asking because `catkin_tools` isn't actively being maintained and I was just curious why you use it and none of the other options. I don't mind what tool MoveIt uses in CI or in the instructions.

Just from a brief look at the instructions I didn't see any special features being used which aren't available in any of the other tools (building, ignoring specific packages, passing CMake build type.

I only noticed that MoveIt use `wstool` for cloning the repositories. As of Noetic the official instructions are using `vcstool` instead - for the same reason that `wstool` is not actively being maintained.

**Comment by gavanderhoorn** [comment]
> I was just asking because `catkin_tools` isn't actively being maintained

it may not be actively maintained, but Colcon is also not recommended for ROS 1 use. At least it wasn't last time I checked.

This leaves us with plain `catkin_make` or `catkin_make_isolated`. Both don't come near the user experience of `catkin_tools`, even if it is unmaintained.

**Comment by facontidavide** [comment]
My two cents: a LOT of people use catkin_tools. One thing is: "we don't want to support any further catkin_tools to focus on something else (Colcon?)", but 0.45 was release 13 months ago and few commits have been done in the midtime.

Can we just have a release to make those changes available to everybody?

**Comment by dirk-thomas** [comment]
> a LOT of people use catkin_tools

That is certainly the case. It doesn't change the fact though that nobody seems to be willing to actively maintain it. Afaik @mikepurvis is currently the only person putting any effort in it.

> Can we just have a release to make those changes available to everybody?

Maybe @mikepurvis can clarify what the pending items are which need to happen before the next release. Then anyone interested in a sooner release can consider to help on those.

**Comment by davetcoleman** [comment]
@dirk-thomas at a high level: we've switched to ``vcstool`` and ``colcon`` for MoveIt 2 (@henningkayser), but I'd very much rather stuck to the classic ROS 1 tools for this last ros 1 release rather than putting a lot of work into migrating into a new tool set. 

Our MoveIt 1 ``master`` branch is also suppose to be backwards compatible with Melodic, as we move to Noetic, another reason to stay with these tools.

**Comment by wjwwood** [comment]
I'm going to try to make a release today or failing that this week at @davetcoleman's request, but I simply don't have the bandwidth to keep this going. So we need someone new to step up if you all are going to continue to use it.

**Comment by davetcoleman** [comment]
Thanks @wjwwood !!

**Comment by wjwwood** [comment]
> So we need someone new to step up if you all are going to continue to use it.

And by this I mean we need people to help @mikepurvis and others, who are still active. I was thinking more the future.

@mikepurvis should I release as-is? I'm going to check into the reported issue with Python 3.8 first, but is there anything else that should block a release?

**Comment by wjwwood** [comment]
I misunderstood the report about 3.8, I see now that it was already fixed on master. Ok, I'll assume master is good to go (I tried it out on 3.8) and release it.

**Comment by wjwwood** [comment]
catkin_tools 0.5.0 is release, but there's still an issue with the deb version. I'm looking into that now... Also, I see a new commit on master, so I'll probably do a 0.5.1 release as soon as I figure out why trollius is still a deb depend.

**Comment by roehling** [comment]
@wjwwood: Is this related to the following error I received:

    Setting up python-catkin-tools (0.5.0-1) ...
      File "/usr/lib/python2.7/dist-packages/catkin_tools/execution/executor.py", line 73
        yield from lock
                 ^
    SyntaxError: invalid syntax
    
    dpkg: error processing package python-catkin-tools (--configure):
     installed python-catkin-tools package post-installation script subprocess returned error exit status 101

**Comment by aroller** [comment]
> @wjwwood: Is this related to the following error I received:
> 
> ```
> Setting up python-catkin-tools (0.5.0-1) ...
>   File "/usr/lib/python2.7/dist-packages/catkin_tools/execution/executor.py", line 73
>     yield from lock
>              ^
> SyntaxError: invalid syntax
> 
> dpkg: error processing package python-catkin-tools (--configure):
>  installed python-catkin-tools package post-installation script subprocess returned error exit status 101
> ```

I just received the same error when installing to a `ros:melodic` docker container on a mac.

**Comment by wjwwood** [comment]
I’m aware of the issues and I’m working on a fix. Sorry for the disruption.

**Comment by moriarty** [comment]
```
Traceback (most recent call last):
  File "/usr/bin/catkin", line 11, in <module>
    load_entry_point('catkin-tools==0.5.0', 'console_scripts', 'catkin')()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 480, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2693, in load_entry_point
    return ep.load()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2324, in load
    return self.resolve()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2330, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/lib/python2.7/dist-packages/catkin_tools/commands/catkin.py", line 28, in <module>
    from catkin_tools.common import is_tty
  File "/usr/lib/python2.7/dist-packages/catkin_tools/common.py", line 24, in <module>
    import asyncio
ImportError: No module named asyncio
```

**Comment by wjwwood** [comment]
I've released 0.6.0 which rolls back to the state of 0.4.5 to fix everyone. I'll make a 0.7.0 release when I'm sure it won't break everything again. It may take a few minutes to propagate.

**Comment by moriarty** [comment]
@wjwwood 0.6.0 was is on ros2 but not ros1

**Comment by moriarty** [comment]
```
apt-cache madison python-catkin-tools
python-catkin-tools |    0.6.0-1 | http://packages.ros.org/ros/ubuntu bionic/main amd64 Packages
python-catkin-tools |    0.6.0-1 | http://packages.ros.org/ros2/ubuntu bionic/main amd64 Packages
```

:+1: thanks @wjwwood

**Comment by pbeeson** [comment]
I see 0.6.0 for ros1, but it won't install the .deb for some reason.  Odd since I've never had problems until it tried installing 0.5.0

```
$ sudo apt-get install -f python-catkin-tools
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 python-catkin-tools : Depends: python2 (< 2.8) but it is not installable
                       Depends: python2 (>= 2.7) but it is not installable
                       Depends: python2:any (>= 2.6.6-7~) but it is not installable
E: Unable to correct problems, you have held broken packages.
```

**Comment by wjwwood** [comment]
Yup, I'm now looking at that. Thanks @pbeeson, sorry for the continued disruption.

**Comment by wjwwood** [comment]
@pbeeson I think we know what needs to change, 0.6.1 is coming soon hopefully:

https://github.com/catkin/catkin_tools/pull/612

**Comment by wjwwood** [comment]
Ok, `0.6.1` is out and can be installed on 18.04 at least:

```
% sudo apt install  python-catkin-tools
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following packages will be upgraded:
  python-catkin-tools
1 upgraded, 0 newly installed, 0 to remove and 464 not upgraded.
Need to get 314 kB of archives.
After this operation, 0 B of additional disk space will be used.
Get:1 http://packages.ros.org/ros/ubuntu bionic/main amd64 python-catkin-tools all 0.6.1-1 [314 kB]
Fetched 314 kB in 0s (791 kB/s)
(Reading database ... 276990 files and directories currently installed.)
Preparing to unpack .../python-catkin-tools_0.6.1-1_all.deb ...
Unpacking python-catkin-tools (0.6.1-1) over (0.4.5-1) ...
Setting up python-catkin-tools (0.6.1-1) ...
```

Please everyone let me know if you are still having issues. I'll work on the 0.7.0 release (which will correctly be Python3 only this time) and try to do that this evening or tomorrow.

Again, my apologies for the disruption, I know it impacted a lot of CI and dev environments.

**Comment by mikepurvis** [comment]
Yes, sorry I missed all this discussion— I should have more clearly marked in the top comment that a 0.5.0 release should have waited on a fix to #597, and only targeted Python 3 and 18.04+.

**Comment by wolfv** [comment]
thanks for the release! Updated it on conda-forge as well. I am really happy to have a working version for python 3 on there https://github.com/conda-forge/catkin_tools-feedstock

**Comment by MatthijsBurgh** [comment]
I still experience problems on 18.04 with `python3-catkin-tools`. It installs `0.5.0-1`.
I get the following error during build:
```
+ catkin build -p1 -j1 --no-status
Traceback (most recent call last):
  File "/usr/bin/catkin", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3254, in <module>
    def _initialize_master_working_set():
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3237, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3266, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 584, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 901, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 787, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'osrf-pycommon>0.1.1' distribution was not found and is required by catkin-tools
```

**Comment by gavanderhoorn** [comment]
According to [status_page/melodic_default.html?q=catkin-tools](http://repositories.ros.org/status_page/melodic_default.html?q=catkin-tools), `0.6.1` is the version in the repositories.

Are you using a mirror?

What is the output of `apt-cache policy python-catkin-tools`?

**Comment by timonegk** [comment]
I guess since catkin_tools is now always python3, only python3-catkin-tools should be released and should replace python-catkin-tools. But a problem is that python3-catkin-tools depends on python3-catkin-pkg which conflicts with python-catkin-pkg which is needed by many other ROS packages. So maybe catkin_pkg must be changed so that the two variants don't conflict.

**Comment by MatthijsBurgh** [comment]
@gavanderhoorn Sorry I meant Noetic, 20.04. But also than one should be `0.6.1`

**Comment by dirk-thomas** [comment]
> So maybe catkin_pkg must be changed so that the two variants don't conflict.

`catkin_pkg` is already being split up into the command line tools (`python-catkin-pkg` and `python3-catkin-pkg` which obviously conflict with each other since they contain the same command line tools just with a different shebang) and the Python modules (`python-catkin-pkg-modules` and `python3-catkin-pkg-modules` which are side-by-side installable).

I assume nobody (I probably should say that) uses `catkin_tools` for its API. Instead users use the CLI. Therefore side-by-side installation of the Python 2 and 3 version of `catkin_tools` doesn't make much sense.

So I would say you should install the Python 2 package on platforms before Ubuntu Focal (where ROS 1 targets Python 2 on) and install the (to-be-released) Python 3 package on platforms starting with Ubuntu Focal (which ROS 1 targets Python 3 on).

**Comment by dirk-thomas** [comment]
> Sorry I meant Noetic, 20.04. But also than one should be 0.6.1

For Ubuntu Focal there will be no Python 2 packages for any ROS tools. The Python 3 version of `catkin_tools` will soon be released - see https://github.com/catkin/catkin_tools/issues/603#issuecomment-638484772

**Comment by Jaeyoung-Lim** [comment]
Same issue with `python3-catkin-tools` 
```
Traceback (most recent call last):
  File "/usr/bin/catkin", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3254, in <module>
    def _initialize_master_working_set():
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3237, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3266, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 584, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 901, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 787, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'osrf-pycommon>0.1.1' distribution was not found and is required by catkin-tools
```

**Comment by wjwwood** [comment]
Yeah, sorry folks may be tomorrow, we're busy with Foxy release stuff.

**Comment by mintar** [comment]
@wjwwood : Just a quick bump: The `'osrf-pycommon>0.1.1' distribution was not found` problem still persists.

**Comment by wjwwood** [comment]
Yeah, the `python3-` version will not work right now you will have to use python2, or install master from pip. I am going to try and release a fixed python3 only version soon, I just need enough contiguous time to make sure I get it right. I'll try to get to it this week.

**Comment by VictorLamoine** [comment]
You can also manually install the missing dependency:
```bash
sudo apt install python3-catkin-lint python3-pip
pip3 install osrf-pycommon
```

**Comment by MatthijsBurgh** [comment]
There has not been any progress in this repo for 27 days. I hope this will change soon.

**Comment by DLu** [comment]
Is there a specific Github issue that tracks the `osrf-pycommon` problem? The best I've found is https://github.com/osrf/osrf_pycommon/pull/68/files

**Comment by Tobias-Fischer** [comment]
Is there an update on this?

**Comment by wjwwood** [comment]
No, sorry, it's still on my "short" list, which is unfortunately not so short. I will try to get something done on it soon.

**Comment by tpet** [comment]
osrf-pycommon can also be installed using `apt*`:
`sudo apt install python3-osrf-pycommon`

**Comment by Timple** [comment]
So basically it is just missing as a dependency of the apt package? I would expect apt to install `python3-osrf-pycommon` in this case.

**Comment by wolfv** [comment]
I am confused on the state of this. 
Is the Python 3 support fundamentally broken? What is missing to make it work? I was under the impression that the required work is already done ....

Can -- whoever knows -- file the appropriate issues?

**Comment by wjwwood** [comment]
https://github.com/catkin/catkin_tools/issues/594#issuecomment-644241926

I just need to do the release of the master branch into a new version but correctly, where correctly means doing all the things to make sure it doesn't override the python2 version on platforms where that matters. Unfortunately this is not something I can delegate as it requires write access to our debian servers (packages.ros.org) and is very fragile.

**Comment by mintar** [comment]
> So basically it is just missing as a dependency of the apt package? I would expect apt to install python3-osrf-pycommon in this case.

Yes, all that's still missing is the apt dependency. If you manually install `python3-osrf-pycommon`, everyhing works:

```bash
sudo apt-get install python3-osrf-pycommon python3-catkin-tools
```

That being said, @wjwwood 's comments sound like it's not as easy as it sounds to add that dependency.

**Comment by Tobias-Fischer** [comment]
Another ping on this issue - catkin_tools has been broken for 9 months now ..

**Comment by ckurtz22** [comment]
New issue on python 3.9, they made Lock not iterable anymore:
```build] Found '23' packages in 0.0 seconds.                                                                                                                                      
[build] Package table is up to date.                                                                                                                                             
[build] An internal error occurred!                                                                                                                                              
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/catkin_tools/verbs/catkin_build/build.py", line 550, in build_isolated_workspace
    all_succeeded = run_until_complete(execute_jobs(
  File "/usr/lib/python3.9/site-packages/catkin_tools/execution/executor.py", line 374, in run_until_complete
    return loop.run_until_complete(coroutine)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/usr/lib/python3.9/site-packages/catkin_tools/execution/executor.py", line 282, in execute_jobs
    job_id, succeeded = yield from done_job_f
  File "/usr/lib/python3.9/site-packages/catkin_tools/execution/executor.py", line 73, in async_job
    yield from lock
TypeError: 'Lock' object is not iterable
```

**Comment by ckurtz22** [comment]
> New issue on python 3.9, they made Lock not iterable anymore:
> 
> ```
> [build] Package table is up to date.                                                                                                                                             
> [build] An internal error occurred!                                                                                                                                              
> Traceback (most recent call last):
>   File "/usr/lib/python3.9/site-packages/catkin_tools/verbs/catkin_build/build.py", line 550, in build_isolated_workspace
>     all_succeeded = run_until_complete(execute_jobs(
>   File "/usr/lib/python3.9/site-packages/catkin_tools/execution/executor.py", line 374, in run_until_complete
>     return loop.run_until_complete(coroutine)
>   File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
>     return future.result()
>   File "/usr/lib/python3.9/site-packages/catkin_tools/execution/executor.py", line 282, in execute_jobs
>     job_id, succeeded = yield from done_job_f
>   File "/usr/lib/python3.9/site-packages/catkin_tools/execution/executor.py", line 73, in async_job
>     yield from lock
> TypeError: 'Lock' object is not iterable
> ```

This is the relevant change https://github.com/python/cpython/commit/90d9ba6ef10af32e8dfe0649789c3a8ccf419e95

**Comment by Rayman** [comment]
@ckurtz22 noetic has Python 3.8 so this issue is not blocking the release of `catkin_tools`

**Comment by ckurtz22** [comment]
> @ckurtz22 noetic has Python 3.8 so this issue should not block the release of `catkin_tools`

This is true, but the previous implementation is still deprecated in 3.8. Also for people using Arch Linux, Arch just got updated to 3.9 so it was broken for them (rest of ROS stuff on Python 3.8 is somewhat broken as well, but still works at least)

**Comment by timonegk** [comment]
More than a year has passed since the creation of this issue – @wjwwood I know that you are busy, but `catkin_tools` is well established in the community and should finally be working with python 3. Is there any way you could free up some time for this issue or can we help somehow?

**Comment by mintar** [comment]
Yeah, I'm afraid @wjwwood has given up on catkin_tools. It's a constant source of grief when trying to switch people to Noetic - the first thing you have to tell them is that you can't even install catkin_tools on Noetic via apt-get, you have to use an arcane set of pip commands.

**Comment by gavanderhoorn** [comment]
> **Edit:** @jspricke just pointed out to me that all that's missing is a proper dependency on `python3-osrf-pycommon`, you'll have to install it manually. So no more need for pip; the following works:
> 
> ```shell
> apt install python3-catkin-tools python3-osrf-pycommon
> ```

I was just going to say: this has been the way to install it for quite some time now.

Has also been on ROS Answers for quite some time now: [catkin build in Ubuntu 20.04 noetic?](https://answers.ros.org/question/353113).

**Comment by Timple** [comment]
So what goes wrong then? Because I can see the entry in requirements.txt

**Comment by gavanderhoorn** [comment]
It's not in [stdeb.cfg](https://github.com/catkin/catkin_tools/blob/1684ffc0afa00241bbfa86b837842788802ec58e/stdeb.cfg) I believe.

**Comment by wjwwood** [comment]
I'm still here, just haven't had time. I know it's hard to believe but it really will take me a full day -ish to do it and I have had a free day in a long time. I have said many times in the past I don't have time to keep maintaining this, but unfortunately this requires mine (or someone with packages.ros.org auth) to do this thing.

@timonegk and @mikepurvis messaged me offline and I'm replying to them now. Maybe we can get something done here soon.

I am truly sorry, but I just don't have much free time ever.

> More than a year has passed since the creation of this issue – @wjwwood I know that you are busy, but catkin_tools is well established in the community and should finally be working with python 3. Is there any way you could free up some time for this issue or can we help somehow?

It would help if someone figured out everything that needs to be done to produce the right debian packages and just tell me what buttons to press. A big part of why I haven't done this yet is that I tried to do it more than a year ago and broke everything, so I need to spend a lot of time experimenting to make sure it get it right. Unfortunately I can't give much more guidance without spending a lot of time on it anyway.

**Comment by timonegk** [comment]
I think there are two possibilities:
1. Drop the python 2 version of `catkin_tools` completely, only provide the python3-catkin-tools package (with `Replaces: python-catkin-tools`). In that case, the package still has to be installable for users of python 2 (i.e. all ROS users up to melodic). The problem are currently two dependencies: `python(3)-catkin-pkg` and `python(3)-osrf-pycommon`. The dependency `catkin-pkg` can be replaced with `catkin-pkg-modules` where both versions can be installed next to each other. For `osrf-pycommon`, there is no such option, `python3-orsf-pycommon` and `python-orsf-pycommon` conflict. I don't know if `python-orsf-pycommon` is actually used by the general public or if this conflict can be ignored. Or the same split as for `catkin-pkg` could be done. In general, however, this option is probably the more difficult one.
2. Keep the current release of `python-catkin-tools` (0.6.1, where the source files are from 0.4.5 which is the last version with python 2 support). Every new version would only be published for python 3. We could still change the dependency `catkin-pkg` to `catkin-pkg-modules` to make it installable for the large majority of people. I don't know how the build system exactly works but there is probably a way to change stdeb's command line flags to include `--with-python2=False`.

I will prepare patches for `stdeb.cfg` for both options tonight.

**Comment by mikepurvis** [comment]
One possibility is that the community could maybe establish a shared catkin-tools PPA, and then the OSRF apt repo could just pull from there, though I don't know what the story is for pypi write access.

**Comment by wjwwood** [comment]
Another thing you have to consider @timonegk is which Ubuntu distributions to target. Which is also configured in `stdeb.cfg`.

If you open a pull request I can at least review it in the next week or so.

> One possibility is that the community could maybe establish a shared catkin-tools PPA, and then the OSRF apt repo could just pull from there, though I don't know what the story is for pypi write access.

That can work too. You (and now @timonegk) both have access to write to pypi for `catkin-tools`.

**Comment by timonegk** [comment]
I opened a pull request: https://github.com/catkin/catkin_tools/pull/648. I decided to go with option 2 because option 1 probably does not make much sense. The PR was tested on several different systems (details in the description) and should hopefully make this release possible.

**Comment by lepalom** [comment]
There's a bug in the Packages file of the distribution of this package that make problems to the users. If you are using Ubuntu Focal and follow these steps, you will have python3-catkin-tools(0.5.0-1), that it's a supported version of Python3.

However, if you are using Debian Buster, and you do `$ apt install python3-catkin-tools` the system will install python3-catkin-tools(0.6.1-1) that is NOT compatible with python3, giving the trollius error.

**Comment by maddypie** [comment]
> Yeah, I'm afraid @wjwwood has given up on catkin_tools. It's a constant source of grief when trying to switch people to Noetic - the first thing you have to tell them is that you can't even install catkin_tools on Noetic via apt-get (*), ~you have to use an arcane set of pip commands~.
> 
> **Edit:** @jspricke just pointed out to me that all that's missing is a proper dependency on `python3-osrf-pycommon`, you'll have to install it manually. So no more need for pip; the following works:
> 
> ```shell
> apt install python3-catkin-tools python3-osrf-pycommon
> ```


Hey, I've been trying to build moveIt from this [official tutorial](https://ros-planning.github.io/moveit_tutorials/doc/getting_started/getting_started.html) but ```rosdep install -y --from-paths . --ignore-src --rosdistro noetic``` but rosdep fails to install some packages:
>ros-noetic-warehouse-ros
ros-noetic-control-msgs
ros-noetic-pcl-ros
ros-noetic-eigen-stl-containers
ros-noetic-tf2-eigen
ros-noetic-srdfdom
ros-noetic-ompl
...

So I tried fetching from ROS Testing repositories by:
```
sudo sh -c 'echo "deb http://packages.ros.org/ros-testing/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
sudo apt update
```
and I get this:
>W: Failed to fetch http://packages.osrfoundation.org/gazebo/ubuntu-stable/dists/focal/InRelease  Temporary failure resolving 'packages.osrfoundation.org'
W: Failed to fetch http://packages.ros.org/ros-testing/ubuntu/dists/focal/InRelease  Temporary failure resolving 'packages.ros.org'
W: Some index files failed to download. They have been ignored, or old ones used instead.

Now, I am not able to ```catkin config --extend /opt/ros/${ROS_DISTRO} --cmake-args -DCMAKE_BUILD_TYPE=Release``` because this raises distribution-not-found error:
>    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'osrf-pycommon>0.1.1' distribution was not found and is required by catkin-tools

I tried doing:
```
sudo apt install python3-osrf-pycommon
```
but, this gives me:
>E: Unable to locate package python3-osrf-pycommon

Same happens on ```sudo apt install python-osrf-pycommon```.

Please guide me. Have I gone wrong somewhere?

**Comment by lepalom** [comment]
@maddypie could you explain what are your distro or setup?

**Comment by maddypie** [comment]
> @maddypie could you explain what are your distro or setup?

Ubuntu 20.04

**Comment by jspricke** [comment]
@maddypie that's unrelated to the issue here, please don't hijack issues.
> Temporary failure resolving 'packages.osrfoundation.org'
There is something broken with your internet, please fix this.
You can ask ROS related questions on https://answers.ros.org/.

**Comment by maddypie** [comment]
> @maddypie that's unrelated to the issue here, please don't hijack issues.
> 
> > Temporary failure resolving 'packages.osrfoundation.org'
> 
> There is something broken with your internet, please fix this.
> 
> You can ask ROS related questions on https://answers.ros.org/.

I am sorry. There was a related discussing happening here:


> Yeah, I'm afraid @wjwwood has given up on catkin_tools. It's a constant source of grief when trying to switch people to Noetic - the first thing you have to tell them is that you can't even install catkin_tools on Noetic via apt-get (*), ~you have to use an arcane set of pip commands~.
> 
> **Edit:** @jspricke just pointed out to me that all that's missing is a proper dependency on `python3-osrf-pycommon`, you'll have to install it manually. So no more need for pip; the following works:
> 
> ```shell
> apt install python3-catkin-tools python3-osrf-pycommon
> ```

it tempted to put my error, since it had to do with exactly the same topic

**Comment by lepalom** [comment]
@maddypie seeing @jspricke comment it's clear that you have some trouble with your connection. I have seen several installations in Ubuntu 20.04 of catkin_tools and just doing:

`apt install python3-catkin-tools python3-osrf-pycommon`

is enough.

**Comment by varunagrawal** [comment]
> > So we need someone new to step up if you all are going to continue to use it.
> 
> And by this I mean we need people to help @mikepurvis and others, who are still active. I was thinking more the future.
> 
> @mikepurvis should I release as-is? I'm going to check into the reported issue with Python 3.8 first, but is there anything else that should block a release?

I am volunteering to help. I can start by doing some PR reviews and we can take it from there?

**Comment by timonegk** [comment]
Hi @varunagrawal, thanks for the offer!
The problem with the release is that currently everything is ready, the binaries are built and tested, but @wjwwood has to release them into the ROS Package Server. I wrote to him and he said he would not have time until the current ROS 2 release slows down. Since the release is this Sunday, I hope that he finds time for catkin_tools very soon.

**Comment by varunagrawal** [comment]
Alright, sounds good. If anything changes, please do let me know.

**Comment by MatthijsBurgh** [comment]
When will this be released on pip and/or apt?

**Comment by wolfv** [comment]
0.7.1 is on conda-forge :) https://github.com/conda-forge/catkin_tools-feedstock/pull/18

**Comment by timonegk** [comment]
Yes, thanks for the reminder. I just released the new version for pip. I've also been trying to get @wjwwood to update the apt package on packages.ros.org, but I have been repeatedly put off for months, so I assume that that's not going to happen anytime soon.
Therefore, I opened a Launchpad PPA where you can get the apt package, if you prefer that to pip. The launchpad is available here: https://launchpad.net/~catkin-tools/+archive/ubuntu/ppa, and it can be added with `sudo add-apt-repository ppa:catkin-tools/ppa`, and then the new version is available with a `sudo apt update && sudo apt install python3-catkin-tools`.
The package version is 0.7.1 instead of 0.7.0 for technical reasons (i.e. I messed up the first upload to launchpad).

**Comment by wjwwood** [comment]
I was trying to release onto ROS's packages.ros.org repository (as apt) and also on PyPi for pip, but I'm getting an error in the `setup.py`:

```
% python3 setup.py sdist bdist_wheel
running sdist
running egg_info
creating catkin_tools.egg-info
writing catkin_tools.egg-info/PKG-INFO
writing dependency_links to catkin_tools.egg-info/dependency_links.txt
writing entry points to catkin_tools.egg-info/entry_points.txt
writing requirements to catkin_tools.egg-info/requires.txt
writing top-level names to catkin_tools.egg-info/top_level.txt
writing manifest file 'catkin_tools.egg-info/SOURCES.txt'
reading manifest file 'catkin_tools.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'catkin_tools.egg-info/SOURCES.txt'
running check
creating catkin_tools-0.7.1
creating catkin_tools-0.7.1/catkin_tools
creating catkin_tools-0.7.1/catkin_tools.egg-info
creating catkin_tools-0.7.1/catkin_tools/commands
creating catkin_tools-0.7.1/catkin_tools/execution
creating catkin_tools-0.7.1/catkin_tools/jobs
creating catkin_tools-0.7.1/catkin_tools/jobs/cmake
creating catkin_tools-0.7.1/catkin_tools/jobs/commands
creating catkin_tools-0.7.1/catkin_tools/notifications
creating catkin_tools-0.7.1/catkin_tools/notifications/resources
creating catkin_tools-0.7.1/catkin_tools/notifications/resources/linux
creating catkin_tools-0.7.1/catkin_tools/notifications/resources/osx
creating catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app
creating catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents
creating catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/MacOS
creating catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources
creating catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/en.lproj
creating catkin_tools-0.7.1/catkin_tools/spaces
creating catkin_tools-0.7.1/catkin_tools/verbs
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_build
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_clean
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_config
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_create
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_env
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_init
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_list
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_locate
creating catkin_tools-0.7.1/catkin_tools/verbs/catkin_profile
creating catkin_tools-0.7.1/completion
copying files to catkin_tools-0.7.1...
copying MANIFEST.in -> catkin_tools-0.7.1
copying README.md -> catkin_tools-0.7.1
copying setup.py -> catkin_tools-0.7.1
copying catkin_tools/__init__.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools/argument_parsing.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools/common.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools/config.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools/context.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools/metadata.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools/resultspace.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools/terminal_color.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools/utils.py -> catkin_tools-0.7.1/catkin_tools
copying catkin_tools.egg-info/PKG-INFO -> catkin_tools-0.7.1/catkin_tools.egg-info
copying catkin_tools.egg-info/SOURCES.txt -> catkin_tools-0.7.1/catkin_tools.egg-info
copying catkin_tools.egg-info/dependency_links.txt -> catkin_tools-0.7.1/catkin_tools.egg-info
copying catkin_tools.egg-info/entry_points.txt -> catkin_tools-0.7.1/catkin_tools.egg-info
copying catkin_tools.egg-info/requires.txt -> catkin_tools-0.7.1/catkin_tools.egg-info
copying catkin_tools.egg-info/top_level.txt -> catkin_tools-0.7.1/catkin_tools.egg-info
copying catkin_tools/commands/__init__.py -> catkin_tools-0.7.1/catkin_tools/commands
copying catkin_tools/commands/catkin.py -> catkin_tools-0.7.1/catkin_tools/commands
copying catkin_tools/execution/__init__.py -> catkin_tools-0.7.1/catkin_tools/execution
copying catkin_tools/execution/controllers.py -> catkin_tools-0.7.1/catkin_tools/execution
copying catkin_tools/execution/events.py -> catkin_tools-0.7.1/catkin_tools/execution
copying catkin_tools/execution/executor.py -> catkin_tools-0.7.1/catkin_tools/execution
copying catkin_tools/execution/io.py -> catkin_tools-0.7.1/catkin_tools/execution
copying catkin_tools/execution/job_server.py -> catkin_tools-0.7.1/catkin_tools/execution
copying catkin_tools/execution/jobs.py -> catkin_tools-0.7.1/catkin_tools/execution
copying catkin_tools/execution/stages.py -> catkin_tools-0.7.1/catkin_tools/execution
copying catkin_tools/jobs/__init__.py -> catkin_tools-0.7.1/catkin_tools/jobs
copying catkin_tools/jobs/catkin.py -> catkin_tools-0.7.1/catkin_tools/jobs
copying catkin_tools/jobs/cmake.py -> catkin_tools-0.7.1/catkin_tools/jobs
copying catkin_tools/jobs/output.py -> catkin_tools-0.7.1/catkin_tools/jobs
copying catkin_tools/jobs/utils.py -> catkin_tools-0.7.1/catkin_tools/jobs
copying catkin_tools/jobs/cmake/python.cmake -> catkin_tools-0.7.1/catkin_tools/jobs/cmake
copying catkin_tools/jobs/cmake/python_install_dir.cmake -> catkin_tools-0.7.1/catkin_tools/jobs/cmake
copying catkin_tools/jobs/commands/__init__.py -> catkin_tools-0.7.1/catkin_tools/jobs/commands
copying catkin_tools/jobs/commands/cmake.py -> catkin_tools-0.7.1/catkin_tools/jobs/commands
copying catkin_tools/jobs/commands/make.py -> catkin_tools-0.7.1/catkin_tools/jobs/commands
copying catkin_tools/notifications/__init__.py -> catkin_tools-0.7.1/catkin_tools/notifications
copying catkin_tools/notifications/impl.py -> catkin_tools-0.7.1/catkin_tools/notifications
copying catkin_tools/notifications/resources/linux/catkin_icon.png -> catkin_tools-0.7.1/catkin_tools/notifications/resources/linux
copying catkin_tools/notifications/resources/linux/catkin_icon_red.png -> catkin_tools-0.7.1/catkin_tools/notifications/resources/linux
copying catkin_tools/notifications/resources/osx/catkin build.app/Contents/Info.plist -> catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents
copying catkin_tools/notifications/resources/osx/catkin build.app/Contents/PkgInfo -> catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents
copying catkin_tools/notifications/resources/osx/catkin build.app/Contents/MacOS/catkin build -> catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/MacOS
copying catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/AppIcon.icns -> catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources
copying catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/catkin_blue_logo_only.icns -> catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources
copying catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/en.lproj/Credits.rtf -> catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/en.lproj
copying catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/en.lproj/InfoPlist.strings -> catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/en.lproj
copying catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/en.lproj/MainMenu.nib -> catkin_tools-0.7.1/catkin_tools/notifications/resources/osx/catkin build.app/Contents/Resources/en.lproj
copying catkin_tools/spaces/__init__.py -> catkin_tools-0.7.1/catkin_tools/spaces
copying catkin_tools/spaces/build.py -> catkin_tools-0.7.1/catkin_tools/spaces
copying catkin_tools/spaces/devel.py -> catkin_tools-0.7.1/catkin_tools/spaces
copying catkin_tools/spaces/install.py -> catkin_tools-0.7.1/catkin_tools/spaces
copying catkin_tools/spaces/log.py -> catkin_tools-0.7.1/catkin_tools/spaces
copying catkin_tools/spaces/source.py -> catkin_tools-0.7.1/catkin_tools/spaces
copying catkin_tools/verbs/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs
copying catkin_tools/verbs/catkin_shell_verbs.bash -> catkin_tools-0.7.1/catkin_tools/verbs
copying catkin_tools/verbs/catkin_build/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_build
copying catkin_tools/verbs/catkin_build/build.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_build
copying catkin_tools/verbs/catkin_build/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_build
copying catkin_tools/verbs/catkin_build/color.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_build
copying catkin_tools/verbs/catkin_clean/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_clean
copying catkin_tools/verbs/catkin_clean/clean.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_clean
copying catkin_tools/verbs/catkin_clean/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_clean
copying catkin_tools/verbs/catkin_clean/color.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_clean
copying catkin_tools/verbs/catkin_config/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_config
copying catkin_tools/verbs/catkin_config/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_config
copying catkin_tools/verbs/catkin_create/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_create
copying catkin_tools/verbs/catkin_create/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_create
copying catkin_tools/verbs/catkin_env/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_env
copying catkin_tools/verbs/catkin_env/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_env
copying catkin_tools/verbs/catkin_init/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_init
copying catkin_tools/verbs/catkin_init/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_init
copying catkin_tools/verbs/catkin_list/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_list
copying catkin_tools/verbs/catkin_list/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_list
copying catkin_tools/verbs/catkin_locate/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_locate
copying catkin_tools/verbs/catkin_locate/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_locate
copying catkin_tools/verbs/catkin_profile/__init__.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_profile
copying catkin_tools/verbs/catkin_profile/cli.py -> catkin_tools-0.7.1/catkin_tools/verbs/catkin_profile
copying completion/_catkin -> catkin_tools-0.7.1/completion
copying completion/catkin.bash -> catkin_tools-0.7.1/completion
Writing catkin_tools-0.7.1/setup.cfg
creating dist
Creating tar archive
removing 'catkin_tools-0.7.1' (and everything under it)
running bdist_wheel
running build
running build_py
creating build
creating build/lib
creating build/lib/catkin_tools
copying catkin_tools/argument_parsing.py -> build/lib/catkin_tools
copying catkin_tools/__init__.py -> build/lib/catkin_tools
copying catkin_tools/resultspace.py -> build/lib/catkin_tools
copying catkin_tools/utils.py -> build/lib/catkin_tools
copying catkin_tools/context.py -> build/lib/catkin_tools
copying catkin_tools/metadata.py -> build/lib/catkin_tools
copying catkin_tools/config.py -> build/lib/catkin_tools
copying catkin_tools/common.py -> build/lib/catkin_tools
copying catkin_tools/terminal_color.py -> build/lib/catkin_tools
creating build/lib/catkin_tools/spaces
copying catkin_tools/spaces/build.py -> build/lib/catkin_tools/spaces
copying catkin_tools/spaces/__init__.py -> build/lib/catkin_tools/spaces
copying catkin_tools/spaces/install.py -> build/lib/catkin_tools/spaces
copying catkin_tools/spaces/devel.py -> build/lib/catkin_tools/spaces
copying catkin_tools/spaces/source.py -> build/lib/catkin_tools/spaces
copying catkin_tools/spaces/log.py -> build/lib/catkin_tools/spaces
creating build/lib/catkin_tools/jobs
copying catkin_tools/jobs/__init__.py -> build/lib/catkin_tools/jobs
copying catkin_tools/jobs/cmake.py -> build/lib/catkin_tools/jobs
copying catkin_tools/jobs/utils.py -> build/lib/catkin_tools/jobs
copying catkin_tools/jobs/catkin.py -> build/lib/catkin_tools/jobs
copying catkin_tools/jobs/output.py -> build/lib/catkin_tools/jobs
creating build/lib/catkin_tools/notifications
copying catkin_tools/notifications/__init__.py -> build/lib/catkin_tools/notifications
copying catkin_tools/notifications/impl.py -> build/lib/catkin_tools/notifications
creating build/lib/catkin_tools/verbs
copying catkin_tools/verbs/__init__.py -> build/lib/catkin_tools/verbs
creating build/lib/catkin_tools/commands
copying catkin_tools/commands/__init__.py -> build/lib/catkin_tools/commands
copying catkin_tools/commands/catkin.py -> build/lib/catkin_tools/commands
creating build/lib/catkin_tools/execution
copying catkin_tools/execution/job_server.py -> build/lib/catkin_tools/execution
copying catkin_tools/execution/__init__.py -> build/lib/catkin_tools/execution
copying catkin_tools/execution/jobs.py -> build/lib/catkin_tools/execution
copying catkin_tools/execution/stages.py -> build/lib/catkin_tools/execution
copying catkin_tools/execution/controllers.py -> build/lib/catkin_tools/execution
copying catkin_tools/execution/io.py -> build/lib/catkin_tools/execution
copying catkin_tools/execution/events.py -> build/lib/catkin_tools/execution
copying catkin_tools/execution/executor.py -> build/lib/catkin_tools/execution
creating build/lib/catkin_tools/jobs/commands
copying catkin_tools/jobs/commands/__init__.py -> build/lib/catkin_tools/jobs/commands
copying catkin_tools/jobs/commands/cmake.py -> build/lib/catkin_tools/jobs/commands
copying catkin_tools/jobs/commands/make.py -> build/lib/catkin_tools/jobs/commands
creating build/lib/catkin_tools/verbs/catkin_locate
copying catkin_tools/verbs/catkin_locate/__init__.py -> build/lib/catkin_tools/verbs/catkin_locate
copying catkin_tools/verbs/catkin_locate/cli.py -> build/lib/catkin_tools/verbs/catkin_locate
creating build/lib/catkin_tools/verbs/catkin_config
copying catkin_tools/verbs/catkin_config/__init__.py -> build/lib/catkin_tools/verbs/catkin_config
copying catkin_tools/verbs/catkin_config/cli.py -> build/lib/catkin_tools/verbs/catkin_config
creating build/lib/catkin_tools/verbs/catkin_profile
copying catkin_tools/verbs/catkin_profile/__init__.py -> build/lib/catkin_tools/verbs/catkin_profile
copying catkin_tools/verbs/catkin_profile/cli.py -> build/lib/catkin_tools/verbs/catkin_profile
creating build/lib/catkin_tools/verbs/catkin_env
copying catkin_tools/verbs/catkin_env/__init__.py -> build/lib/catkin_tools/verbs/catkin_env
copying catkin_tools/verbs/catkin_env/cli.py -> build/lib/catkin_tools/verbs/catkin_env
creating build/lib/catkin_tools/verbs/catkin_init
copying catkin_tools/verbs/catkin_init/__init__.py -> build/lib/catkin_tools/verbs/catkin_init
copying catkin_tools/verbs/catkin_init/cli.py -> build/lib/catkin_tools/verbs/catkin_init
creating build/lib/catkin_tools/verbs/catkin_list
copying catkin_tools/verbs/catkin_list/__init__.py -> build/lib/catkin_tools/verbs/catkin_list
copying catkin_tools/verbs/catkin_list/cli.py -> build/lib/catkin_tools/verbs/catkin_list
creating build/lib/catkin_tools/verbs/catkin_clean
copying catkin_tools/verbs/catkin_clean/__init__.py -> build/lib/catkin_tools/verbs/catkin_clean
copying catkin_tools/verbs/catkin_clean/color.py -> build/lib/catkin_tools/verbs/catkin_clean
copying catkin_tools/verbs/catkin_clean/clean.py -> build/lib/catkin_tools/verbs/catkin_clean
copying catkin_tools/verbs/catkin_clean/cli.py -> build/lib/catkin_tools/verbs/catkin_clean
creating build/lib/catkin_tools/verbs/catkin_create
copying catkin_tools/verbs/catkin_create/__init__.py -> build/lib/catkin_tools/verbs/catkin_create
copying catkin_tools/verbs/catkin_create/cli.py -> build/lib/catkin_tools/verbs/catkin_create
creating build/lib/catkin_tools/verbs/catkin_build
copying catkin_tools/verbs/catkin_build/build.py -> build/lib/catkin_tools/verbs/catkin_build
copying catkin_tools/verbs/catkin_build/__init__.py -> build/lib/catkin_tools/verbs/catkin_build
copying catkin_tools/verbs/catkin_build/color.py -> build/lib/catkin_tools/verbs/catkin_build
copying catkin_tools/verbs/catkin_build/cli.py -> build/lib/catkin_tools/verbs/catkin_build
creating build/lib/catkin_tools/jobs/cmake
copying catkin_tools/jobs/cmake/python.cmake -> build/lib/catkin_tools/jobs/cmake
Traceback (most recent call last):
  File "setup.py", line 144, in <module>
    cmdclass={'install': PermissiveInstall},
  File "/usr/local/lib/python3.6/dist-packages/setuptools/__init__.py", line 145, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/local/lib/python3.6/dist-packages/wheel/bdist_wheel.py", line 192, in run
    self.run_command('build')
  File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/lib/python3.6/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/local/lib/python3.6/dist-packages/setuptools/command/build_py.py", line 53, in run
    self.build_package_data()
  File "/usr/local/lib/python3.6/dist-packages/setuptools/command/build_py.py", line 126, in build_package_data
    srcfile in self.distribution.convert_2to3_doctests):
TypeError: argument of type 'NoneType' is not iterable
```

@timonegk did you run into this at all? I'm using this setup:

```
% lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.5 LTS
Release:	18.04
Codename:	bionic

% python3 --version
Python 3.6.9

% python3 -c 'import setuptools; print(setuptools.__version__)'
40.8.0
```

**Comment by tykurtz** [comment]
Jumping to (maybe) help move this along.

Would the following docker image help?

```dockerfile
FROM ubuntu:18.04

RUN apt-get update && apt-get install -y \
    build-essential                      \
    wget                                 \
    python3-dev                          \
    python3-setuptools                   \
    python3-wheel                        \
 && rm -rf /var/lib/apt/lists/*

RUN wget https://github.com/catkin/catkin_tools/archive/refs/tags/0.7.1.tar.gz && \
    tar xf 0.7.1.tar.gz && \
    rm -rf 0.7.1.tar.gz

RUN cd catkin_tools-0.7.1 && \
    /usr/bin/python3.6 setup.py sdist bdist_wheel
```

I couldn't reproduce your issue, but you should be able to build the artifact.

**Comment by jspricke** [comment]
> I was trying to release onto ROS's packages.ros.org repository (as apt) and also on PyPi for pip, but I'm getting an error in the `setup.py`:
> 
> ```
> % python3 setup.py sdist bdist_wheel
[...]
> copying catkin_tools/jobs/cmake/python.cmake -> build/lib/catkin_tools/jobs/cmake
> Traceback (most recent call last):
>   File "setup.py", line 144, in <module>
>     cmdclass={'install': PermissiveInstall},
>   File "/usr/local/lib/python3.6/dist-packages/setuptools/__init__.py", line 145, in setup
>     return distutils.core.setup(**attrs)
>   File "/usr/lib/python3.6/distutils/core.py", line 148, in setup
>     dist.run_commands()
>   File "/usr/lib/python3.6/distutils/dist.py", line 955, in run_commands
>     self.run_command(cmd)
>   File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
>     cmd_obj.run()
>   File "/usr/local/lib/python3.6/dist-packages/wheel/bdist_wheel.py", line 192, in run
>     self.run_command('build')
>   File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
>     self.distribution.run_command(command)
>   File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
>     cmd_obj.run()
>   File "/usr/lib/python3.6/distutils/command/build.py", line 135, in run
>     self.run_command(cmd_name)
>   File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
>     self.distribution.run_command(command)
>   File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
>     cmd_obj.run()
>   File "/usr/local/lib/python3.6/dist-packages/setuptools/command/build_py.py", line 53, in run
>     self.build_package_data()
>   File "/usr/local/lib/python3.6/dist-packages/setuptools/command/build_py.py", line 126, in build_package_data
>     srcfile in self.distribution.convert_2to3_doctests):
> TypeError: argument of type 'NoneType' is not iterable
> ```

@wjwwood note the `/usr/local/python3.6` in the traceback, you have a modified Python installed breaking the command.
Using the Ubuntu Python it works fine, as Tyler reported.

**Comment by wjwwood** [comment]
Actually I don't have a custom python installation. But I do have an updated setuptools and usr local lib is the right place for non-apt dependencies. It might be an issue that it doesn't work with newer versions of setuptools which is something I don't want to ignore. It's recommended to use the latest version when releasing to pypi. Thanks for the suggestions though, I'll look into it more tomorrow.

**Comment by timonegk** [comment]
I did not run into this problem. (And I did already successfully release the package on PyPi.)

**Comment by wjwwood** [comment]
>  The package version is 0.7.1 instead of 0.7.0 for technical reasons (i.e. I messed up the first upload to launchpad).

@timonegk you probably already realize this, but next time you release, do at least `0.7.3` otherwise your workaround will prevent you from putting the "real" `0.7.2` on the PPA. I thought there was a `-N` you could append to the version to solve issues like this (packaging issues), because using a future version number causes conflicts with the source repository.

> It might be an issue that it doesn't work with newer versions of setuptools which is something I don't want to ignore.

In case anyone else ran into this problem...

It turned out to be a bug in the version of `setuptools` I was using. I narrowed it down to a few changes, but in the end just updating `setuptools` fixed it.

I was surprised at this issue (and I still think it has to do with something "special" we're doing in our `setup.py`) because I had just released `osrf_pycommon` version `1.0.0` which also is the first Python3-only release of that, and I didn't have this problem.

---

0.7.1 should now be available via packages.ros.org too. I really am sorry it took so long, but finally I got the Python3 only version sufficiently tested and released for `osrf_pycommon` which was a prerequisite for this last week. And I spent some time this weekend convincing myself that I wasn't going to disrupt existing uses of out catkin tools apt packages, like I accidentally did last time.

From now on, updating the version should be quick (now that the Python3 only is not new), so just ping me on an issue if you want a new one @timonegk. I actually had quite a few small issues to resolve when releasing the debs for `osrf_pycommon` version `1.0.0`, so it wasn't very straightforward the first time, but I don't expect problems in the future.

I will also still follow up with someone about maybe just pulling from your PPA, rather than us making and uploading debs. That becomes even easier for me, since all we will do is run a build farm job on demand after you update the PPA to pull in new stuff. But I can do manual releases until then.

---

Despite the testing, if issues come up, please let me know and I'll fix it as quick as I can (now that it's live).

**Comment by wjwwood** [comment]
> The package version is 0.7.1 instead of 0.7.0 for technical reasons (i.e. I messed up the first upload to launchpad).

Actually I misunderstood you, I see you've already fixed that by doing a 0.7.1 release locally. Sorry, my bad :)

Still I think, in the future, you may be able to fix that by putting something like `0.7.0-2` on the PPA.

**Comment by jspricke** [comment]
> Still I think, in the future, you may be able to fix that by putting something like `0.7.0-2` on the PPA.

@wjwwood A `-2` can only change files in the `debian/` directory. If I got it correct, the actual fix was in the `stdeb.cfg` which is part of the upstream project: https://github.com/catkin/catkin_tools/pull/648/commits/7a8c239a19096f1b6a829179d82b26e9e3dbff37, so a `0.7.1` was the right thing to do.

**Comment by wjwwood** [comment]
Hmm, the `stdeb.cfg` is only used as part of the release and isn't installed. so I would have thought that only affects stuff in the debian folder, but maybe not.

**Comment by jspricke** [comment]
@wjwwood 'installed' does not matter, point is what is in the source package and where is it versioned. I just had a look at the PPA and the `stdeb.cfg` is indeed not part of the source package but as the `debian/` directory is generated from it, it was still the right way to make the package reproducible.

**Comment by timonegk** [comment]
I uploaded 0.7.0 with the wrong stdeb.cfg, and after the fix I was not able to upload it with the same version, so I had to increase it to 0.7.1.

**Comment by wjwwood** [comment]
Sorry, I didn't mean to imply that they way @timonegk did it was wrong, that's certainly not the case. I agree having the tagged version be consistent with what is released is a good thing. And version numbers are essentially free.

I was just pointing out that if you need to do so (just information that might be useful) you can adjust the "debian version", `stdeb` has this option: https://github.com/astraw/stdeb/blob/3e6522cefa240be365ebdc0dc0c7923ba7118999/stdeb/util.py#L84 and therefore `0.7.0-1` could become `0.7.0-2` if the mistake was purely packaging related. That would let you upload to the ppa without a new release. Like if you accidentally used an old version of stdeb the first time and the generated `debian/` folder was wrong but you didn't notice until you uploaded it, or some other packaging only issue came up.

It's kind of peculiar (compared to other packaging processes) that we include this packaging information in the source repository. But I agree that keeping the `stdeb.cfg` correct in the source repository and tagged with the release, since we do include it.

**Comment by mikepurvis** [comment]
Having the packaging info in-source I think kind of implies that it should be a `3.0 (native)` rather than `3.0 (quilt)` format package, eg a version number with no debian increment at all. But I expect that that would be even weirder to tools like stdeb than just including the increment and rarely/never using it.
